FirebaseSessions/Sources/FirebaseSessions.swift

```diff
@@ -89,23 +98,104 @@ protocol SessionsProvider {
-      if self.settings.sessionsEnabled, session.shouldDispatchEvents {
-        let event = SessionStartEvent(sessionInfo: session, appInfo: self.appInfo)
-        DispatchQueue.global().async {
-          self.coordinator.attemptLoggingSessionStart(event: event) { result in
-          }
+      // Generating a Session ID early is important as Subscriber
+      // SDKs will need to read it immediately upon registration.
+      let sessionInfo = self.sessionGenerator.generateNewSession()
+
+      // Post a notification so subscriber SDKs can get an updated Session ID
+      self.notificationCenter.post(name: Sessions.SessionIDChangedNotificationName,
+                                   object: nil)
+
+      let event = SessionStartEvent(sessionInfo: sessionInfo, appInfo: self.appInfo)
+
+      // Wait until all subscriber promises have been fulfilled before
+      // doing any data collection.
+      all(self.subscriberPromises.values).then(on: .global(qos: .background)) { _ in
+        guard self.isAnyDataCollectionEnabled else {
+          Logger
+            .logDebug(
+              "Data Collection is disabled for all subscribers. Skipping this Session Event"
+            )
+          return
+        }
+
+        Logger.logDebug("Data Collection is enabled for at least one Subscriber")
+
+        // Fetch settings if they have expired. This must happen after the check for
+        // data collection because it uses the network, but it must happen before the
+        // check for sessionsEnabled from Settings because otherwise we would permanently
+        // turn off the Sessions SDK when we disabled it.
+        self.settings.updateSettings()
+
+        self.addEventDataCollectionState(event: event)
+
+        if !(self.settings.sessionsEnabled && sessionInfo.shouldDispatchEvents) {
+          Logger
+            .logDebug(
+              "Session Event logging is disabled sessionsEnabled: \(self.settings.sessionsEnabled), shouldDispatchEvents: \(sessionInfo.shouldDispatchEvents)"
+            )
+          return
+        }
+
+        self.coordinator.attemptLoggingSessionStart(event: event) { result in
         }
-      } else {
-        Logger.logDebug("Session logging is disabled by configuration settings.")
       }
     }
   }
 
+  // MARK: - Data Collection
+
+  var isAnyDataCollectionEnabled: Bool {
+    for subscriber in subscribers {
+      if subscriber.isDataCollectionEnabled {
+        return true
+      }
+    }
+    return false
+  }
+
+  func addEventDataCollectionState(event: SessionStartEvent) {
+    subscribers.forEach { subscriber in
+      event.set(subscriber: subscriber.sessionsSubscriberName,
+                isDataCollectionEnabled: subscriber.isDataCollectionEnabled)
+    }
+  }
+
+  // MARK: - SessionsProvider
+
+  var currentSessionPayload: SessionPayload {
+    return SessionPayload(sessionId: sessionGenerator.currentSession?.sessionId ?? "")
+  }
+
+  func register(subscriber: SessionsSubscriber) {
```

Comment:
Would there be an option for the dependencies to unregister? I think we should provide the API to unsubscribe from hearing updates.

Reply:
I don't see a need to yet - Crash and Perf all live for the lifetime of the app

Reply:
Fair point. But have a thought around how the unregister will look like. Infact, if the product data collection flags toggle, in the future, we might want to honor that by avoiding transactions across SDKs. So, an unregister would be required in that context (not immediately).

Reply:
Sounds good - maybe we can add this in a follow up

Reply:
Sounds good.